README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# MetSizeR

MetSizeR is a Shiny application developed for metabolomic scientists to
estimate the optimal sample size required for a study to achieve a
desired statistical power for their study, even in cases where pilot
data are not available. The MetSizeR application is developed based on
the methodology by Nyamundanda et al. (2013).

## Installation

MetSizeR can be installed from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("MetSizeR")
```

## Accessing MetSizeR application

All MetSizeR functionality is accessed through a Shiny application,
which can be accessed by running the following command:

``` r
# load MetSizeR package
library(MetSizeR)

# launch MetSizeR Shiny application
MetSizeR()
```

## References

Nyamundanda, G., Gormley, I. C., Fan, Y., Gallagher, W. M., & Brennan,
L. (2013). MetSizeR: Selecting the optimal sample size for metabolomic
studies using an analysis based approach. BMC Bioinformatics, 14(1),
338. <https://doi.org/10.1186/1471-2105-14-338>


NEWS.md
# MetSizeR 2.0.0

* Removed previous package GUI which relied on dependency 'gWidgets' which is no longer maintained.
* Moved package functionality to a Shiny application.
* Improved code performance by updating operations to use 'Rfast' package methods where possible.
* Added functionality to handle categorical covariates.
* Increased data sizes accepted by the MetSizeR algorithm.
* Restructured package code for readability - separated functions for different purposes.
* Added a `NEWS.md` file to track changes to the package.


R/utils.R
#' Unit Scale Covariates
#'
#' @param C Matrix containing covariates.
#'
#' @return Matrix containing covariates whose values are scaled to lie between 0 and 1.
#'
#' @noRd
standardize <- function(C) {
  for (i in 1:ncol(C))
  {
    rg <- range(C[, i])
    C[, i] <- (C[, i] - min(C[, i])) / (rg[2] - rg[1])
  }
  C
}


#' Sample from the null distribution for test statistics
#'
#' @param T Number of permutations to estimate for each sampling distribution.
#' @param S Matrix of NA values of size p times T where p is the number of
#'   metabolites and T is the number of permutations to estimate for each
#'   sampling distribution. Use in function is as a storage matrix for standard
#'   deviation with correction factor added.
#' @param TS Matrix of NA values of size p times T where p is the number of
#'   metabolites and T is the number of permutations to estimate for each
#'   sampling distribution. Use in function is as a storage matrix for test
#'   statistics.
#' @param x Matrix of simulated data.
#' @param y Vector containing group membership labels for samples. Should
#'   contain entry 1 for samples in group 1 and entry 2 for samples in group 2,
#'   for a total of n1star + n2star entries, where n1star and n2star are the
#'   group sample sizes at the current sample size iteration.
#' @param n11 The current group 1 sample size minus 1.
#' @param n22 The current group 2 sample size minus 1.
#' @param in1n2 The sum of the reciprocals of current group 1 sample size and
#'   the current group 2 sample size.
#' @param nn2 The sum of the current group 1 sample size and
#'   the current group 2 sample size, minus 2.
#' @param cpcf The 5th percentile of the number of bins.
#'
#' @return List containing matrix S and matrix TS. Matrix S contains the
#'   pooled standard deviations with correction factor added. Matrix TS contains the
#'   test statistics for the designated test - a two-sample test statistic used
#'   to test a difference in the means of two groups.
#'
#' @noRd
samp.dist <- function(T, S, TS, x, y, n11, n22, in1n2, nn2, cpcf) {
  for (t in 1:T)
  {
    yperm <- sample(y, replace = FALSE)
    x1 <- x[yperm == 1, ]
    x2 <- x[yperm == 2, ]
    # sd
    Sj <- sqrt(in1n2 * (n11 * Rfast::colVars(x1) + n22 * Rfast::colVars(x2)) / nn2)
    # stored sd with correction factor
    S[, t] <- Sj + Rfast::Sort(Sj, partial = cpcf)[cpcf]
    TS[, t] <- (Rfast::colmeans(x1) - Rfast::colmeans(x2)) / S[, t]
  }
  return(list(S = S, TS = TS))
}


R/plotting.R
#' Plot metsize results
#'
#' @param plot_data Dataframe containing results_sim data from metsize function.
#' @param FDR The target FDR for the experiment.
#' @param nhat The estimate optimal sample size for the experiment.
#'
#' @return Plot object containing plot of FDR vs sample size for metsize data.
#'
#' @noRd
plot_fun <- function(plot_data, FDR, nhat) {
  # aes_ and ~variable_name  used here to pass CRAN checks
  ggplot2::ggplot(
    plot_data,
    ggplot2::aes_(x = ~sample_size, y = ~fdr_50_percentile)
  ) +
    ggplot2::geom_point(colour = 4, size = 2) +
    ggplot2::geom_line(colour = 4, size = 1) +
    # percentile lines
    ggplot2::geom_ribbon(
      ggplot2::aes_(ymin = ~fdr_10_percentile, ymax = ~fdr_90_percentile),
      linetype = 2,
      alpha = 0.1,
      colour = 4,
      size = 1
    ) +
    # target fdr line
    ggplot2::geom_hline(yintercept = FDR, linetype = 3, size = 1) +
    # result line
    ggplot2::geom_vline(
      ggplot2::aes_(
        xintercept = nhat[1],
        linetype = paste(
          "Group 1: ",
          nhat[2],
          "\nGroup 2: ",
          nhat[3]
        )
      ),
      colour = "darkblue",
      size = 1,
      key_glyph = "blank"
    ) +
    ggplot2::labs(
      x = "Sample Size",
      y = "FDR",
      linetype = paste("   Estimated Optimal Sample Size: ", nhat[1], "  ")
    ) +
    ggplot2::annotate(
      geom = "text",
      x = min(plot_data$sample_size),
      y = FDR,
      vjust = -1,
      label = paste("FDR = ", FDR)
    ) +
    ggplot2::theme(
      legend.position = c(0.95, 0.95),
      legend.justification = c("right", "top"),
      legend.key = ggplot2::element_blank(),
      legend.title = ggplot2::element_text(size = 12),
      legend.text = ggplot2::element_text(size = 12),
      legend.key.size = ggplot2::unit(0, "lines"),
      plot.title = ggplot2::element_text(hjust = 0.5, size = 22),
      axis.title = ggplot2::element_text(size = 12)
    )
}


R/data_simulation.R
#' Simulate data using pilot data
#'
#' Simulate data for use in the MetsizeR algorithm using parameters gleaned from
#' available pilot data for the experiment
#'
#' @param n1 Group 1 minimum sample size to test.
#' @param n2 Group 2 minimum sample size to test.
#' @param p Number of bins.
#' @param Zerop Vector of zeros of length p.
#' @param Ip Identity matrix of dimension p.
#' @param Zeroq Vector of zeros where length is the number of dimensions of the latent space.
#' @param Iq Identity matrix with same dimensions as the latent space.
#' @param ZeroL Vector of zeros where length is the number of covariates present. For PPCCA only.
#' @param IL Identity matrix where dimension is the number of covariates present. For PPCCA only.
#' @param ZeroL1 Vector of zeros where length is the number of covariates present, plus 1. For PPCCA only.
#' @param IL1 Identity matrix where dimension is the number of covariates present, plus 1. For PPCCA only.
#' @param sig Posterior mode estimate of the variance of the error terms of the model.
#' @param W Loadings matrix.
#' @param Alpha Maximum likelihood estimates of the regression coefficients associated with the covariates in the PPCCA model. For PPCCA only.
#' @param mu Mean vector for pilot data metabolites.
#' @param mod Vector of possible models.
#' @param model Chosen model.
#'
#' @return Matrix of simulated data.
#'
#' @noRd
sim.pilot.data <- function(n1, n2, p, Zerop, Ip, Zeroq, Iq, ZeroL, IL, ZeroL1, IL1, sig, W, Alpha, mu, mod, model) {
  # generate parameters
  n <- n1 + n2
  if (mod[1] == model) {  # PPCA
    u <- Rfast::rmvnorm(n, Zeroq, Iq)
  } else {  # PPCCA
    C <- Rfast::rmvnorm(n, ZeroL, IL)
    # Standardize covariates for stability
    C <- standardize(C)
    C <- rbind(rep(1, n), t(C))
    u <- Rfast::rmvnorm(n, Zeroq, Iq) + t(Alpha %*% C)
  }
  # generate data
  x <- Rfast::rmvnorm(n, Zerop, sig * Ip) + tcrossprod(u, W) + matrix(mu, n, p, byrow = TRUE)  # keep tcrossprod in base because faster for needed dimensions
  return(x)
}

#' Simulate data without using pilot data
#'
#' Simulate data for use in the MetsizeR algorithm when pilot data are not
#' available. Parameters are gleaned from the expected structure of the data.
#'
#' @param n1 Group 1 minimum sample size to test.
#' @param n2 Group 2 minimum sample size to test.
#' @param p Number of bins.
#' @param Zerop Vector of zeros of length p.
#' @param Ip Identity matrix of dimension p.
#' @param q Number of dimensions of the latent space.
#' @param Zeroq Vector of zeros where length is the number of dimensions of the latent space.
#' @param Iq Identity matrix with same dimensions as the latent space.
#' @param ZeroL Vector of zeros where length is the number of covariates present. For PPCCA only.
#' @param IL Identity matrix where dimension is the number of covariates present. For PPCCA only.
#' @param ZeroL1 Vector of zeros where length is the number of covariates present, plus 1. For PPCCA only.
#' @param IL1 Identity matrix where dimension is the number of covariates present, plus 1. For PPCCA only.
#' @param alpha.sigma Scale parameter of the prior distribution of the variance.
#' @param beta.sigma Shape parameter of the prior distribution of the variance.
#' @param mod Vector of possible models.
#' @param model Chosen model.
#' @param ao1 A vector of length q where each entry is the scale parameter of the prior distribution of the variance.
#' @param bo A vector containing c(0.5*(ao-1),0.25*(ao-1)).
#'
#' @return Matrix of simulated data.
#'
#' @noRd
sim.pilot <- function(n1, n2, p, Zerop, Ip, q, Zeroq, Iq, ZeroL, IL, ZeroL1, IL1, alpha.sigma, beta.sigma, mod, model, ao1, bo) {
  # generate parameters
  n <- n1 + n2
  sig <- 1 / stats::rgamma(1, alpha.sigma, beta.sigma)
  if (mod[1] == model) {  # PPCA
    u <- Rfast::rmvnorm(n, Zeroq, Iq)
  } else {  # PPCCA
    Alpha <- Rfast::rmvnorm(q, ZeroL1, 3 * IL1)
    C <- Rfast::rmvnorm(n, ZeroL, IL)
    # Standardize covariates for stability
    C <- standardize(C)
    C <- rbind(rep(1, n), t(C))
    u <- Rfast::rmvnorm(n, Zeroq, Iq) + t(Alpha %*% C)
  }
  v <- 1 / stats::rgamma(q, ao1, bo)
  W <- Rfast::rmvnorm(p, Zeroq, v * Iq)
  # generate data
  x <- Rfast::rmvnorm(n, Zerop, sig * Ip) + tcrossprod(u, W)
  return(x)
}


R/vary_proportions.R
#' Runs metsize function for specified proportions of significant bins.
#'
#' @param props A vector containing the proportions of significant bins on which to run metsize funtion.
#' @param p The number of bins.
#' @param fdr The target FDR.
#'
#' @return A list containing the metsize results for each of the input proportions.
#'
#' @noRd
metsize_props <- function(props, p, fdr, mod, num_covs = 0) {
  props <- as.vector(stats::na.omit(props))

  results_list <- vector(mode = "list", length = length(props))

  # run metsize function for each proportion given and store results
  for (i in 1:length(props)) {
    result <- metsize(pilot = NULL, n1 = 4, n2 = 4, p = p, prop = props[i], covars = NULL, ncovar = num_covs, model = mod, plot.prop = FALSE, target.fdr = fdr, Targeted = FALSE)

    results_list[[i]] <- list(results_sim = result$results_sim, nhat = result$nhat)
  }

  return(results_list)
}


R/metsize.R
#' Metsize
#'
#' Function to estimate sample size required to achieve a desired statistical
#' power.
#'
#' @param pilot A matrix containing pilot data, if available.
#' @param n1 Sample group 1 minimum sample size to consider.
#' @param n2 Sample group 2 minimum sample size to consider.
#' @param p Number of bins in the data.
#' @param prop Expected proportion of significant bins.
#' @param covars Matrix of covariates, if available. Categorical covariates
#'   should be included as X-1 dummy variables for a covariate with X levels.
#' @param ncovar The number of covariates present. For Z numeric covariates and Y categorical covariates with X total levels, this should be Z + X - Y.
#' @param model The chosen model.
#' @param plot.prop TRUE/FALSE should sample size be calculated for varying proportions of significant bins?
#' @param target.fdr The desired false discovery rate.
#' @param Targeted TRUE/FALSE is the experiment intended as targeted analysis?
#'
#' @return List containing the estimated sample size with the per-group sample
#'   size breakdown, results of FDR calculations for each sample size tested,
#'   results for varying the proportion of significant metabolites, the number
#'   of bins input, the number of covariates input, the model type used, the
#'   input minimum sample sizes for group 1 and group 2, and the desired FDR.
#'
#' @export
#'
#' @references G. Nyamundanda, I. C. Gormley, Y. Fan, W. M. Gallagher and L. Brennan, BMC Bioinformatics, 2013, 14, 338.
#'
#' @examples
#' metsize(pilot = NULL, n1 = 4, n2 = 4, p = 200, prop = 0.2, covars = NULL,
#'     ncovar = 0, model = "PPCA", plot.prop = FALSE, target.fdr = 0.10,
#'     Targeted = FALSE)
#' metsize(pilot = NULL, n1 = 6, n2 = 4, p = 100, prop = 0.1, covars = NULL,
#'     ncovar = 2, model = "PPCCA", plot.prop = FALSE, target.fdr = 0.05,
#'     Targeted = TRUE)
metsize <- function(pilot = pilot, n1 = 4, n2 = 4, p = 200, prop = 0.25, covars = covars, ncovar = 0, model = "PPCA", plot.prop = FALSE, target.fdr = 0.05, Targeted = FALSE) {
  mod <- c("PPCA", "PPCCA")
  # get relevant parameters from pilot data
  if (length(pilot) != 0) {
    if (mod[2] == model) { # ppcca
      ncovar <- ncol(covars)
      resppcca <- MetabolAnalyze::ppcca.metabol(pilot, covars, 2, 2, scale = "unit")
      W <- resppcca$loadings
      sig <- resppcca$sig
      Alpha <- resppcca$coefficients
      mu <- colMeans(pilot)
      p <- ncol(pilot) # gets number of bins/metabs from pilot data
    } else { # ppca
      pilot <- sweep(pilot, 2, colMeans(pilot), "-")
      resppca <- MetabolAnalyze::ppca.metabol(pilot, 2, 2, scale = "unit")
      W <- resppca$loadings
      sig <- resppca$sig
      mu <- colMeans(pilot)
      p <- ncol(pilot) # gets number of bins/metabs from pilot data
    }
  }

  if (plot.prop == FALSE) {
    mprop <- c(prop) # proportion of significant metabolites
    mrange <- ceiling(mprop * p) # different number significant metabolites
    mtry <- length(mrange) # number of different m values
    nprop.increment <- 1 # Sample size increments
    sfactors <- (n1 + n1) * nprop.increment
  } else { # vestigal proportion plotting
    mprop <- c(prop, seq(0.1, 0.5, 0.1)) # proportion of significant metabolites
    mrange <- ceiling(mprop * p) # different number significant metabo
    mtry <- length(mrange) # number of different m values
    nprop.increment <- c(1, 2, 4, 7) # Sample size increments
    sfactors <- (n1 + n2) * nprop.increment # four different sample sizes to be considered
  }

  # Setting up the initial values and the prior parameters
  T <- 20 # number of permutations to estimate the sampling distribution.
  Sim <- 20 # Number of pilot data sets simulated.
  L <- ncovar # number of covariates
  n <- n1 + n2 # sample size for pilot data
  q <- 2 # dimension of the latent space
  alpha.sigma <- ao <- 5 # the scale parameter of the prior distribution of the variance.
  beta.sigma <- 2 * (alpha.sigma - 1) # the shape parameter of the prior distribution of the variance.
  bo <- c(0.5 * (ao - 1), 0.25 * (ao - 1))
  ao1 <- rep(ao, q)
  n.increment <- c(1:8) # Sample size increments considered.
  ntry <- length(n.increment) # number of different samples sizes to be considered.

  # Storing statistics for each sample size
  fdr50_sim <- matrix(NA, ntry, 1)
  fdr90_sim <- matrix(NA, ntry, 1)
  fdr10_sim <- matrix(NA, ntry, 1)
  fdr_sim <- matrix(NA, ntry, Sim)

  # Storing statistics for each m values
  fdr50_prop <- matrix(NA, mtry - 1, length(sfactors))
  fdr90_prop <- matrix(NA, mtry - 1, length(sfactors))
  fdr10_prop <- matrix(NA, mtry - 1, length(sfactors))
  Add <- array(NA, c(p, T, ntry))
  TSstore <- array(NA, c(p, T, ntry))

  # define effect size
  # Based on variance of underlying simulation model
  delta <- stats::qnorm(0.99)
  if (length(pilot) != 0) {
    delta <- stats::qnorm(0.89)
  }

  # objects for data simulation and storage
  cf <- 0.05 # for correction factor
  Zeroq <- rep(0, q)
  Iq <- diag(q) # no need to change to Rfast because small dims, q = 2
  Zerop <- rep(0, p)
  Ip <- Rfast::Diag.matrix(p, v = 1)
  if (mod[2] == model) {
    ZeroL <- rep(0, L)
    IL <- diag(L) # no need to change to Rfast because small dims, L < 10 typically
    ZeroL1 <- rep(0, (L + 1))
    IL1 <- diag(L + 1) # no need to change to Rfast because small dims, L < 10 typically
  }
  TS <- S <- matrix(NA, p, T)
  cpcf <- ceiling(p * cf) # for correction factor

  # for each prop
  for (m in 1:mtry) {
    pstat <- 1 - (mrange[m] / p)
    ind <- matrix(FALSE, p, T)
    pos <- sample(1:p, size = mrange[m]) # sampling mrange[m] metabolites
    ind[pos, ] <- TRUE # matrix indicating truly significant and non significant metabolites
    i <- 0 # counter

    # for each sample size
    for (k in 1:ntry)
    {
      n1star <- n1 * n.increment[k] # sample size for treatment group 1
      n2star <- n2 * n.increment[k] # sample size for treatment group 2
      # further objects needed
      nstar <- n1star + n2star
      in1n2 <- 1 / n1star + 1 / n2star
      n11 <- n1star - 1
      n22 <- n2star - 1
      nn2 <- n1star + n2star - 2
      Add.sd <- sqrt(in1n2)

      y <- c(rep(1, n1star), rep(2, n2star))

      # If influence of different proportion of significant metabolites is not of interest
      if (m == 1) {

        # assessing the effect of repeated simulations from the underlying model
        for (s in 1:Sim) {
          # Simulating the pilot data
          if (length(pilot) != 0) {
            x <- sim.pilot.data(n1star, n2star, p, Zerop, Ip, Zeroq, Iq, ZeroL, IL, ZeroL1, IL1, sig, W, Alpha, mu, mod, model)
          } else {
            x <- sim.pilot(n1star, n2star, p, Zerop, Ip, q, Zeroq, Iq, ZeroL, IL, ZeroL1, IL1, alpha.sigma, beta.sigma, mod, model, ao1, bo)
          }

          # Estimating the sampling distribution of the test statistic using permutations
          res.sampdist <- samp.dist(T, S, TS, x, y, n11, n22, in1n2, nn2, cpcf)
          TS <- res.sampdist$TS
          S <- res.sampdist$S

          # Store the test statistics for the current sample size
          TSstore[, , k] <- TS

          # calculating the shift in metabolites
          vars <- S / Add.sd
          Add[, , k] <- delta / (vars * Add.sd)

          # estimating the FDR
          tsB <- TS
          # Add an increment factor to the test statistics of the truly significant metabolites
          tsB[pos, ] <- tsB[pos, ] + Add[pos, , k]
          atsB <- abs(tsB)
          # identifying a cut-off point (m-th largest absolute value of the p TsB values)
          crit <- stats::quantile(atsB, pstat)
          # false discovery rate calculation
          errors <- (colSums(atsB > crit & !ind)) / (colSums(atsB > crit))
          fdr_sim[k, s] <- stats::quantile(errors[!is.na(errors)], 0.5) # median FDR of the Tstar permutations
        }

        # assessing the effect of repeated simulations from the underlying model
        emp <- stats::quantile(fdr_sim[k, ], c(0.1, 0.5, 0.9))
        fdr10_sim[k] <- emp[1]
        fdr50_sim[k] <- emp[2]
        fdr90_sim[k] <- emp[3]
      } else {
        # assessing the effect of varying the proportion of truly significant metabolites (increasing m values) on four different sample sizes.
        if (any(nprop.increment == n.increment[k])) {
          i <- i + 1
          # adding a shift component to truly significant metabolites
          tsB <- TSstore[, , k]
          tsB[pos, ] <- tsB[pos, ] + Add[pos, , k]
          atsB <- abs(tsB)
          crit <- stats::quantile(atsB, 1 - (mrange[m] / p))
          # calculate FDR
          errors <- (colSums(atsB > crit & !ind)) / (colSums(atsB > crit))
          emp <- stats::quantile(errors[!is.na(errors)], c(0.1, 0.5, 0.9))
          fdr10_prop[m - 1, i] <- emp[1]
          fdr50_prop[m - 1, i] <- emp[2] # median FDR for Tstar permutations permutations
          fdr90_prop[m - 1, i] <- emp[3]
        } # if(any)
      } # if
    } # k
  } # m

  # store results and format nicely
  results_sim <- cbind(n.increment * n, fdr50_sim, fdr90_sim, fdr10_sim)
  results_prop <- cbind(mprop[-1], fdr50_prop, fdr10_prop, fdr90_prop)
  colnames(results_sim) <- c("sample_size", "fdr_50_percentile", "fdr_90_percentile", "fdr_10_percentile")
  colnames(results_prop) <- c("prop", paste("fdr50_prop", sfactors, sep = "_"), paste("fdr10_prop", sfactors, sep = "_"), paste("fdr90_prop", sfactors, sep = "_"))
  results_prop <- list(results_prop, sfactors)
  names(results_prop) <- c("results_prop", "sample_sizes")

  # Determine the sample size at which the FDR line is equal to 0.05
  # by linear regression
  opty <- rep(0, 2)
  ind1 <- min(c(1:nrow(results_sim))[results_sim[, 2] < target.fdr])
  ind2 <- max(c(1:ind1)[results_sim[1:ind1, 2] > target.fdr])
  opty <- c(results_sim[ind1, 2], results_sim[ind2, 2])
  optx <- c(results_sim[ind1, 1], results_sim[ind2, 1])
  optres <- stats::lm(opty ~ optx)
  nhat <- round((target.fdr - optres$coef[1]) / optres$coef[2])

  # calculate returned values based on input sample sizes for groups
  if (n1 == n2) {
    if (Rfast::is_integer(nhat / 2)) {
      n1 <- n2 <- nhat / 2
    } else {
      nhat <- nhat + 1
      n1 <- n2 <- nhat / 2
    }
  } else {
    if (Rfast::is_integer(n1 * nhat / (n1 + n2))) {
      n1 <- n1 * nhat / (n1 + n2)
      n2 <- nhat - n1
    } else {
      n1.user <- n1
      n2.user <- n2
      n1 <- ceiling(n1.user * nhat / (n1.user + n2.user))
      n2 <- ceiling(n2.user * nhat / (n1.user + n2.user))
      nhat <- n1 + n2
    }
  }
  est <- c(nhat, n1, n2)
  names(est) <- c("n", "n1", "n2")

  return(list(nhat = est, results_sim = results_sim, results_prop = results_prop, p = p, prop = prop, ncovar = ncovar, model = model, n1 = n1, n2 = n2, target.fdr = target.fdr))
}
